tools/wptrunner/wptrunner/executors/base.py

```diff
@@ -805,3 +806,36 @@ def process_action(self, url, payload):
 
     def _send_message(self, message_type, status, message=None):
         self.protocol.testdriver.send_message(message_type, status, message=message)
+
+
+class ActionContext(object):
+    def __init__(self, logger, protocol, context):
+        self.logger = logger
+        self.protocol = protocol
+        self.context = context
+        self.initial_window = None
+        self.switched_frame = False
+
+    def __enter__(self):
+        if self.context is None:
+            return
+
+        window = self.context[0]
+        if window:
+            self.initial_window = self.protocol.base.current_window
+            self.logger.debug("Switching to window %s" % window)
+            self.protocol.testdriver.switch_to_window(window)
+
+        for frame_id in self.context[1:]:
+            self.switched_frame = True
+            self.logger.debug("Switching to frame %s" % frame_id)
+            self.protocol.testdriver.switch_to_frame(frame_id)
+
+    def __exit__(self, *args):
+        if self.initial_window is not None:
+            self.logger.debug("Switching back to initial window")
+            self.protocol.base.set_window(self.initial_window)
+            self.initial_window = None
+        elif self.switched_frame:
+            self.protocol.testdriver.switch_to_frame(None)
```

Comment:
Can we get here? I don't quite follow how `window` (`self.context[0]`) can be falsy in the code as written; isn't it always set if context is non-null?

(Or is there a situation where `self.protocol.base.current_window` could have been `None`?)